### PR TITLE
doc: update the doc of plugin customize

### LINKED ISSUE #3840: [Feature][Metrics] Allow users to import data and expand data schema by CSV
### Search before asking

- [X] I had searched in the [issues](https://github.com/apache/incubator-devlake/issues?q=is%3Aissue) and found no similar feature requirement.

### Use case

As a data analyst, I want to measure the engineering metrics and group them by different dimensions, for example:
- the `commit count` by internal and outsourced employees (`accounts` dimension)
- the `commit count` by work-types such as bug, feat, docs, etc. (`commits` dimension)
- the `issue lead time` of internal requirements and customer-requested requirements (`issues` dimension)
- the `issue lead time` of issues in different hierarchies of business lines (`issues` dimension)

However, I don't have the raw data in DevLake for calculation. There're 2 cases:

- Case 1: I use a [DevLake-supported data source](https://github.com/apache/incubator-devlake#-supported-data-sources), Jira or GitHub for instance. 
   In this case, I have the [issues](https://devlake.apache.org/docs/DataModels/DevLakeDomainLayerSchema#issues), [accounts](https://devlake.apache.org/docs/DataModels/DevLakeDomainLayerSchema#accounts), [issue_commits](https://devlake.apache.org/docs/DataModels/DevLakeDomainLayerSchema#issue_commits), and [commits](https://devlake.apache.org/docs/DataModels/DevLakeDomainLayerSchema#commits) data collected by DevLake.  I only need to add other dimensional data that is not in the [domain layer tables](https://devlake.apache.org/docs/DataModels/DevLakeDomainLayerSchema).
- Case 2: I use a data source that is not supported by DevLake, ClickUp and Trello for instance. (More basic and higher priority compare to Case 1)
   In this case, I have to import both dimensional data and the data in domain layer tables such as [issues](https://devlake.apache.org/docs/DataModels/DevLakeDomainLayerSchema#issues), [accounts](https://devlake.apache.org/docs/DataModels/DevLakeDomainLayerSchema#accounts), and [issue_commits](https://devlake.apache.org/docs/DataModels/DevLakeDomainLayerSchema#issue_commits). (`commits` can always be collected by DevLake's [GitExtractor](https://devlake.apache.org/docs/Plugins/gitextractor)).

In either case, the dimensional data is highly customizable, as different companies use different dimensions to analyze. Hence, users should be able to:
- customize schema: add new fields or tables to the domain layer tables
- import customized data: add their own raw/dimensional data to the updated tables

### Proposed Solutions

Here are a few solutions to meet the above use cases.

#### Solution 1 - Customize schema and import data via `.csv`
1. Send a doc to data analysts when they enquire how to do custom analysis based on their own dimensional data.
   a. This doc lists three templates, `issues.csv`, `accounts.csv`, and `commit_issues`, and introduces how to use them.
   b. Each of the above templates contains several fields. For instance, `issues.csv` contains fields (* are required fields):
   - issue_id *
   - issue_key *
   - issue_original_status
   - issue_resolution_date
   - story_points
   - priority
   - severity
   - component
   - epic_key
   - assignee_name
   - custom_field_1
   - custom_field_2
   - ...
   - custom_field_10
2. Guide users to download these templates in the terminal
   a. In case 1, users download the templates with data **queried from** the corresponding entities.
   b. In case 2, users download the **blank** templates.
3. Users fill in the templates
   a. In case 1, users add values to custom fields and rename these fields.
   b. In case 2, users add all data to the blank templates and rename these fields
4. Users upload the .csv in the terminal. Data gets imported or updated.

**Question for solution 1:**
1. From a schema design perspective, which is the better option?
   - add custom fields as new columns in `issues` and `accounts`
   - add the custom fields to a new dimensional table for each entity, such as `issue_dimensions`, `account_dimensions`, `commit_dimensions`
2. In case 2, shall users manually import distinct issue_ids? This might be too complicated for end users.
3. How will the data schema change if the user upgrades DevLake?

#### Solution 2 - Customize schema and import data via Config UI's Blueprint
1. After users set up DevLake, users create a blueprint and go to Jira, GitHub, and TAPD connections' 'set transformation' page
2. Users add custom_fields (just like adding multiple GitHub tokens) to the domain layer tables, such as `custom_field_1`, `custom_field_2`, ..., to`custom_field_10`.
3. In each custom field, users can select the raw layer keys to get their values transformed into the domain layer.

In short, solution 2 is a GUI alternative to solution 1 with a better user experience. My idea is to implement solution 1 first.

#### To Do (If solution 1 is adopted)

- [ ] The APIs to upload the .csv templates for `issues`, `commits`, `issue_commits` and `accounts`
   - [ ] define the templates for these entities
   - [ ] determine the CRUD strategy for these data
- [ ] add a field `original_project_name` to table.issues. The value of this field comes from Jira project name, TAPD workspace name, or GitHub's user/repo)
- [ ] Add a doc for customized research/adding analytical dimensions
- [ ] Re-organize doc links in README and website to improve the observability of the manual; otherwise, data analysts won't find it

#### .CSV Templates

**issue.csv**
 - issue_id *
 - issue_key *
 - issue_original_status
 - issue_resolution_date
 - story_points
 - priority
 - severity
 - component
 - epic_key
 - assignee_name
 - custom_field_1
 - custom_field_2
 - ...
 - custom_field_10

**commits.csv**
- commit_sha *
- authored_email * 
- authored_date *
- commit_msg * 
- dev_eq *
- repo_id *
- custom_field_1
- custom_field_2
- ...
- custom_field_10

**issue_commits**
- issue_id
- commit_sha

**accounts**
- account_id *:  primary email
- name *
- email_list
- custom_field_1
- custom_field_2
- ...
- custom_field_10

### Related issues

_No response_

### Are you willing to submit a PR?

- [ ] Yes I am willing to submit a PR!

### Code of Conduct

- [X] I agree to follow this project's [Code of Conduct](https://www.apache.org/foundation/policies/conduct)

### PR DESCRIPTION
**Pull request by mindlesscloud**
# Summary
Update the document of plugin customize.
The documentation of the old API has been updated and provided new documents for the new APIs.
### Does this close any open issues?
Closes [#3840](https://github.com/apache/incubator-devlake/issues/3840)

### CODE REVIEW

File: docs/Plugins/customize.md [hunk 1]
```diff
@@ -11,10 +11,17 @@ description: >
 This plugin provides users the ability to create/delete columns and extract data from certain raw layer tables.
 The columns created with this plugin must be start with the prefix `x_`
 
-**NOTE:** All columns created by this plugin are of the datatype `VARCHAR(255)`
+**NOTE:** For now, only the following 5 types were supported:
```
Comment (Startrekzky) [inline docs/Plugins/customize.md]: Text change: For now, only the following five types are supported:

File: docs/Plugins/customize.md [hunk 2]
```diff
@@ -11,10 +11,17 @@ description: >
 This plugin provides users the ability to create/delete columns and extract data from certain raw layer tables.
```
Comment (Startrekzky) [inline docs/Plugins/customize.md]: This plugin provides users the ability to: 
- Add/delete columns in domain layer tables[URL]
- Insert values to the new columns with data extracted from certain raw layer tables

File: docs/Plugins/customize.md [hunk 3]
```diff
@@ -11,10 +11,17 @@ description: >
...
 The columns created with this plugin must be start with the prefix `x_`
```
Comment (Startrekzky) [inline docs/Plugins/customize.md]: The names of columns added via this plugin must start with the prefix `x_`

File: docs/Plugins/customize.md [hunk 4]
```diff
@@ -75,25 +82,129 @@ curl 'http://localhost:8080/pipelines' \
...
+
 > DELETE /plugins/customize/issues/fields/x_test
+
+### Upload `issues.csv` file
+
+> POST /plugins/customize/csvfiles/issues.csv
+
+The HTTP `Content-Type` must be  `multipart/form-data`, and the form should have three fields:
+
+- `file` the CSV file
```
Comment (Startrekzky) [inline docs/Plugins/customize.md]: `file`: The CSV file

File: docs/Plugins/customize.md [hunk 5]
```diff
@@ -75,25 +82,129 @@ curl 'http://localhost:8080/pipelines' \
...
+DevLake will parse the CSV file and store it in the `issues` table, where the `labels` are stored in the `issue_labels` table. 
+If the `boardId` does not appear, a new record will be created in the boards table. The `board_issues` table will be updated at the same time as the import.
+The following is an issues.CSV file sample:
+
+|id                           |created_at             |updated_at             |_raw_data_params     |_raw_data_table          |_raw_data_id|_raw_data_remark|url                                                                 |icon_url|issue_key|title        |description                      |epic_key|type |status|original_status|story_point|resolution_date|created_date                 |updated_date                 |parent_issue_id|priority|original_estimate_minutes|time_spent_minutes|time_remaining_minutes|creator_id                                           |creator_name|assignee_id                                          |assignee_name|severity|component|lead_time_minutes|original_project|original_type|x_int         |x_time             |x_varchar|x_float|labels              |
```
Comment (Startrekzky) [inline docs/Plugins/customize.md]: I reckon that `created_at` and `updated_at` are system fields in MySQL, it should not be defined by users?
image

File: docs/Plugins/customize.md [hunk 6]
```diff
@@ -75,25 +82,129 @@ curl 'http://localhost:8080/pipelines' \
...
+|bitbucket:BitbucketIssue:1:14|2022-09-15 15:27:56.395|2022-09-15 15:27:56.395|board789             |_raw_bitbucket_api_issues|49          |                |https://api.bitbucket.org/2.0/repositories/thenicetgp/lake/issues/14|        |14       |issue test011|issue test011                    |        |issue|TODO  |new            |0          |NULL           |2022-08-12 13:45:12.810+00:00|2022-08-12 13:45:12.810+00:00|               |blocker |0                        |0                 |0                     |bitbucket:BitbucketAccount:1:62abf394192edb006fa0e8cf|tgp         |bitbucket:BitbucketAccount:1:62abf394192edb006fa0e8cf|tgp          |        |         |NULL             |NULL            |NULL         |41534568464351|2022-09-15 15:27:56|NULL     |NULL   |label1,label2,label3|
+
+
+### Upload `issue_commitss.csv` file
+
+> POST /plugins/customize/csvfiles/issue_commits.csv
+
+The `Content-Type` should be  `multipart/form-data`, and the form should have three fields:
+
+- `file` the CSV file
```
Comment (Startrekzky) [inline docs/Plugins/customize.md]: These are not the three fields to be imported.

File: docs/Plugins/customize.md [hunk 7]
```diff
@@ -75,25 +82,129 @@ curl 'http://localhost:8080/pipelines' \
...
+|bitbucket:BitbucketIssue:1:14|2022-09-15 15:27:56.395|2022-09-15 15:27:56.395|board789             |_raw_bitbucket_api_issues|49          |                |https://api.bitbucket.org/2.0/repositories/thenicetgp/lake/issues/14|        |14       |issue test011|issue test011                    |        |issue|TODO  |new            |0          |NULL           |2022-08-12 13:45:12.810+00:00|2022-08-12 13:45:12.810+00:00|               |blocker |0                        |0                 |0                     |bitbucket:BitbucketAccount:1:62abf394192edb006fa0e8cf|tgp         |bitbucket:BitbucketAccount:1:62abf394192edb006fa0e8cf|tgp          |        |         |NULL             |NULL            |NULL         |41534568464351|2022-09-15 15:27:56|NULL     |NULL   |label1,label2,label3|
+
+
+### Upload `issue_commitss.csv` file
```
Comment (Startrekzky) [inline docs/Plugins/customize.md]: Typo: issue_commits.csv

File: docs/Plugins/customize.md [hunk 8]
```diff
@@ -75,25 +82,129 @@ curl 'http://localhost:8080/pipelines' \
...
+    "displayName": "varchar",
+    "dataType": "varchar(255)",
+    "description": "test for varchar"
+  }
 ]
+
 ```
-Create extension column `x_test` for the table `issues`
+
+### Create A Customized Column
```
Comment (Startrekzky) [inline docs/Plugins/customize.md]: “Create a Customized Column”. Use lowercase for crowns like a, an, the

**Review by Startrekzky — CHANGES_REQUESTED** [review]

File: docs/Plugins/customize.md (hunk 5 above)
Comment (mindlesscloud) [reply docs/Plugins/customize.md]: No, it was set by the code. However, usually, the orm framework will help us set them up automatically

File: docs/Plugins/customize.md (hunk 2 above)
Comment (mindlesscloud) [reply docs/Plugins/customize.md]: Not only the new columns but all the columns could get their new values in this way. I will change the text to:

Insert values to certain columns with data extracted from some raw layer tables

Does it sound good to you?

**Review by Startrekzky — APPROVED** [review]
LGTM

**Pull request merged**